src/ddspdrum/module.py

```diff
@@ -368,16 +368,21 @@ def _npyforward(self, mod_signal: np.array, phase: float = 0.0) -> np.ndarray:
 
         """
 
-        assert (mod_signal >= 0).all() and (mod_signal <= 1).all()
+        assert (mod_signal >= -1).all() and (mod_signal <= 1).all()
+
+        control_as_frequency = self.make_control_as_frequency(mod_signal)
 
-        modulation = self.p("mod_depth") * mod_signal
-        control_as_midi = self.p("pitch") + modulation
-        control_as_frequency = midi_to_hz(control_as_midi)
         cosine_argument = self.make_argument(control_as_frequency) + phase
 
+        # Store final phase.
         self.phase = cosine_argument[-1]
         return self.oscillator(cosine_argument)
 
+    def make_control_as_frequency(self, mod_signal: np.ndarray):
+        modulation = self.p("mod_depth") * mod_signal
+        control_as_midi = self.p("pitch") + modulation
+        return midi_to_hz(control_as_midi)
+
```

Comment:
Factoring this section, as this is what `FmVCO` has to replace. Could be a better name for this method.

Effectively we're pivoting between various representations of instructions: (1) *midi* values, (2) *instantaneous frequency* (Hz), and (3) *phase*, which ultimately controls the oscillator.

Reply:
pitch_signal_to_frequency? I think what you have is also okay